### PR TITLE
fix(api): refactor controller code

### PR DESCRIPTION
**Pull request by michellemtchai**
- Please merge this PR after PR #220 
- create reusable helper function in application_helper
- fix redundant declaration of route in routes.rb

### CODE REVIEW

**Review by dingtaoliu — COMMENTED** [review]
Could you please add some documentation for the helper functions?  I’m just worried that in the future when someone wants to modify an existing controller action, they will have a hard time because the helper functions are really generalized.

File: api/app/helpers/application_helper.rb [hunk 1]
```diff
@@ -11,4 +11,74 @@ def json(data, include: nil, except: nil)
...
+            render_success(custom_table_data ? table : table.order(:id))
+            return
+        end
+        return if invalid_id(filter_table, filter_id)
+
+        render_success(filter_func)
+    end
+
+    def table_to_id(table)
+        "#{table.to_s.gsub(/([a-z])([A-Z])/, '\1_\2').downcase}_id".to_sym
```
Comment (siefkenj) [inline api/app/helpers/application_helper.rb]: This really scares me...I don't like auto conversion. I would rather have a dictionary lookup.

File: api/app/helpers/application_helper.rb [hunk 2]
```diff
@@ -11,4 +11,74 @@ def json(data, include: nil, except: nil)
         end
         data
     end
+
+    def filter_given_id(table, filter_id, custom_table_data = true)
```
Comment (siefkenj) [inline api/app/helpers/application_helper.rb]: Add documentation comments to each of these functions explaining what they do.

File: api/app/helpers/application_helper.rb [hunk 3]
```diff
@@ -11,4 +11,74 @@ def json(data, include: nil, except: nil)
...
+        table_data = custom_table_data ? table : table.order(:id)
+        table_data.select do |entry|
+            entry[filter_id] == params[filter_id].to_i
+        end
+    end
+
+    def index_response(table, filter_table, filter_func, custom_table_data = false)
```
Comment (siefkenj) [inline api/app/helpers/application_helper.rb]: You're passing in an active record, not a table, right?

File: api/app/helpers/application_helper.rb [hunk 4]
```diff
@@ -11,4 +11,74 @@ def json(data, include: nil, except: nil)
...
+        if entry.update_attributes!(param_reqs) && result
+            update_fn&.call(entry)
+            render_success(parts_fn ? merge_fn.call(entry) : entry)
+        else
+            full_error_render(parts_fn ? { errors: errors } : entry)
+        end
+    end
+
+    def delete_entry(table, destroy_sub_parts = nil)
+        params.require(:id)
```
Comment (siefkenj) [inline api/app/helpers/application_helper.rb:146]: Don't depend on global variables. If you need `params`, pass them in.

File: api/app/helpers/application_helper.rb [hunk 5]
```diff
@@ -11,4 +11,74 @@ def json(data, include: nil, except: nil)
...
+            else
+                render_success(output ? output.call : entry)
+            end
+        else
+            entry.destroy!
+            full_error_render(entry, output&.call)
+        end
+    end
+
+    def update_condition(table)
```
Comment (siefkenj) [inline api/app/helpers/application_helper.rb]: This is not a good name

File: api/app/controllers/api/v1/applicants_controller.rb [hunk 6]
```diff
@@ -5,46 +5,22 @@ module Api::V1
...
-            end
-            return if invalid_id(Session, :session_id)
-
-            render_success(applicants_by_session)
+            index_response(Applicant, Session, applicants_by_session)
         end
 
         # POST /applicants
         def create
-            # if we passed in an id that exists, we want to update
```
Comment (siefkenj) [inline api/app/controllers/api/v1/applicants_controller.rb:19]: Don't delete this comment

File: api/app/helpers/application_helper.rb (hunk 3 above)
Comment (siefkenj) [inline api/app/helpers/application_helper.rb]: You're passing in an active record, not a table, right?

**Comment by siefkenj** [comment]
Replaced by PR #257

**Pull request closed without merging**